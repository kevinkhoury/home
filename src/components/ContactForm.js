import React, { useState } from 'react'
import { v4 as uuidv4 } from 'uuid'
import _ from 'lodash'
import PropTypes from 'prop-types'
import fetch from 'node-fetch'

function encode(data) {
  return Object.keys(data)
    .map((key) => encodeURIComponent(key) + '=' + encodeURIComponent(data[key]))
    .join('&')
}

const responseMessages = {
  initial: null,
  working: null,
  success: 'Thank you! Your message has been sent.',
  error: 'Whoops, something went wrongâ€¦ Please try again.',
}

const ContactForm = ({ formText, isPreview }) => {
  const botfield = uuidv4()
  const [values, setValues] = useState({
    [botfield]: '',
    name: '',
    email: '',
    message: '',
    checkbox: true,
  })
  const [sendStatus, updateSendStatus] = useState('initial')

  const btnClasses = [
    'button',
    'large',
    sendStatus === 'working' ? 'disabled' : '',
  ]
    .filter((item) => item)
    .join(' ')
  const btnText = sendStatus === 'success' ? 'Success!' : formText.submit

  const handleChange = (e) => {
    const vals = _.cloneDeep(values)
    const { name, type, value } = e.target
    vals[name] = type === 'checkbox' ? !vals[name] : value
    setValues(vals)
  }

  const handleSubmit = (e) => {
    e.preventDefault()
    if (isPreview) return null
    if (!!values[botfield]) return null
    updateSendStatus('working')
    const form = e.target
    fetch('/', {
      method: 'POST',
      headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
      body: encode({
        'form-name': form.getAttribute('name'),
        ...values,
      }),
    })
      .then((result) => {
        const success = result && result.status && result.status === 200
        updateSendStatus(success ? 'success' : 'error')
      })
      .catch(() => {
        updateSendStatus('error')
      })
  }
  return (
    <div className="contact-form-holder">
      <form
        name="contact"
        method="post"
        data-netlify="true"
        data-netlify-honeypot={botfield}
        onSubmit={handleSubmit}
        className={`contact-form ${sendStatus}`}
      >
        {/* The `form-name` hidden field is required to support form submissions without JavaScript */}
        <input type="hidden" name="form-name" value="contact" />
        <div hidden>
          <input
            name={botfield}
            placeholder={botfield}
            onChange={handleChange}
            tabIndex="-1"
            aria-label={botfield}
          />
        </div>
        <input
          className="input"
          type={'text'}
          name={'name'}
          placeholder={formText.name}
          onChange={handleChange}
          id={'name'}
          value={values.name}
          required={true}
          aria-label={formText.name}
        />
        <input
          className="input"
          type={'email'}
          name={'email'}
          placeholder={formText.email}
          onChange={handleChange}
          id={'email'}
          value={values.email}
          required={true}
          aria-label={formText.email}
        />
        <textarea
          className="textarea"
          name={'message'}
          placeholder={formText.message}
          onChange={handleChange}
          id={'message'}
          value={values.message}
          required={true}
          rows={4}
          aria-label={formText.message}
        />
        <button className={btnClasses} type="submit">
          {btnText}
        </button>
      </form>
      <div className={`response-msg ${sendStatus}`}>
        {responseMessages[sendStatus]}
      </div>
    </div>
  )
}

ContactForm.propTypes = {
  formText: PropTypes.shape({
    name: PropTypes.string.isRequired,
    email: PropTypes.string.isRequired,
    message: PropTypes.string.isRequired,
    submit: PropTypes.string.isRequired,
  }).isRequired,
  isPreview: PropTypes.bool,
}

export default ContactForm
