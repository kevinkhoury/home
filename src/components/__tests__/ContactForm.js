import React from 'react'
import renderer from 'react-test-renderer'
import fetch from 'node-fetch'
import { render, fireEvent, screen, waitFor } from '@testing-library/react'
import ContactForm from '../ContactForm'

const { Response } = jest.requireActual('node-fetch')

const DEFAULT_PROPS = {
  formText: {
    name: 'name',
    email: 'email',
    message: 'message',
    submit: 'submit',
  },
  isPreview: false,
}

jest.mock('node-fetch')

jest.mock('uuid', () => ({
  v4: jest.fn(() => 'uuid'),
}))

afterEach(() => {
  jest.clearAllMocks()
})

describe('ContactForm', () => {
  it('snapshot renders properly', () => {
    const tree = renderer.create(<ContactForm {...DEFAULT_PROPS} />)
    expect(tree.toJSON()).toMatchSnapshot()
  })
  it('encodes form values into a URI string', () => {
    const encode = ContactForm.__get__('encode')
    const result = encode({
      foo: 'foo',
      bar: 'baz',
    })
    expect(result).toBe('foo=foo&bar=baz')
  })
  it('updates form text on user input', () => {
    render(<ContactForm {...DEFAULT_PROPS} />)
    const input = screen.getByPlaceholderText('name')
    fireEvent.change(input, { target: { value: 'foobar' } })
    expect(input.value).toBe('foobar')
  })
  it("doesn't submit on preview", () => {
    const props = { ...DEFAULT_PROPS, isPreview: true }
    render(<ContactForm {...props} />)
    fireEvent.click(screen.getByRole('button', { name: /submit/i }))
    expect(fetch).not.toHaveBeenCalled()
  })
  it("doesn't submit if the honeypot has text", () => {
    render(<ContactForm {...DEFAULT_PROPS} />)
    const botfield = screen.getByPlaceholderText('uuid')
    fireEvent.change(botfield, { target: { value: 'foobar' } })
    fireEvent.click(screen.getByRole('button', { name: /submit/i }))
    expect(fetch).not.toHaveBeenCalled()
  })
  it('succeeds on normal submit', async () => {
    fetch.mockReturnValueOnce(Promise.resolve(new Response('foobar')))
    render(<ContactForm {...DEFAULT_PROPS} />)
    fireEvent.click(screen.getByRole('button', { name: /submit/i }))
    await waitFor(() => expect(screen.getByText(/Thank you/)).toBeTruthy())
  })
  it('fails on network error', async () => {
    fetch.mockReturnValueOnce(Promise.reject('not working'))
    render(<ContactForm {...DEFAULT_PROPS} />)
    fireEvent.click(screen.getByRole('button', { name: /submit/i }))
    await waitFor(() => expect(screen.getByText(/Whoops/)).toBeTruthy())
  })
})
