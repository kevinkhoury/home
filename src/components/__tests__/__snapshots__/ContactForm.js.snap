// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactForm snapshot renders properly 1`] = `
<div
  className="contact-form-holder"
>
  <form
    className="contact-form initial"
    data-netlify="true"
    data-netlify-honeypot="uuid"
    method="post"
    name="contact"
    onSubmit={[Function]}
  >
    <input
      name="form-name"
      type="hidden"
      value="contact"
    />
    <div
      hidden={true}
    >
      <input
        aria-label="uuid"
        name="uuid"
        onChange={[Function]}
        placeholder="uuid"
        tabIndex="-1"
      />
    </div>
    <input
      aria-label="name"
      className="input"
      id="name"
      name="name"
      onChange={[Function]}
      placeholder="name"
      required={true}
      type="text"
      value=""
    />
    <input
      aria-label="email"
      className="input"
      id="email"
      name="email"
      onChange={[Function]}
      placeholder="email"
      required={true}
      type="email"
      value=""
    />
    <textarea
      aria-label="message"
      className="textarea"
      id="message"
      name="message"
      onChange={[Function]}
      placeholder="message"
      required={true}
      rows={4}
      value=""
    />
    <button
      className="button large"
      type="submit"
    >
      submit
    </button>
  </form>
  <div
    className="response-msg initial"
  />
</div>
`;
